### ISSUE TITLE
Plan for CoreCLR and Mono to run corefx tests in their CI

### ISSUE THREAD

**Issue by safern** [dotnet/corefx#34957]
I've opened this issue to have a discussion on what would be the best plan to follow regarding on CoreFX delivering test assets to Mono and CoreCLR.

I know that @mikem8361 has been doing some work to publish tests assets to an azure blob storage with a well known name, and then CoreCLR would restore that. However there has been some discussion on why using blob storage and not NuGet packages.

I know that @ericstj had some points against using NuGet packages, regarding the package size and the layout of the package as it will not be following any NuGet conventions.

cc: @marek-safar @danmosemsft @jkotas @RussKeldorph

**Comment by mikem8361** [comment]
My plan so far. I already have the corefx uploading steps prototyped.

* Change corefx the builds to add a step to upload the test asset zips (around 270) to blob storage. It only takes 30-50 seconds to upload these zip file per "config".
    * Using Arcade’s Microsoft.DotNet.Build.Tasks.Feed/PushToBlobFeed build tasks.
    * Give the blob storage a strongly named like: corefx-tests/$(OfficialBuildId)/$(OSGroup).$(ArchGroup).$(ConfigurationGroup)/$(TargetGroup).
    * Using the dotnet-core blob storage account that already has a build def variable group with the PAT.
* Upload the corefx binaries to this storage URL too as a zip file (basically most of the current corefx Helix test job correlation payload).  We also could just use the nuget packages that are published for the build, but that would really be painful to use on the coreclr test side.
* Optionally change the corefx Helix test step to use this blob storage account by sending the URL of the test asset zips instead of being part of the Helix payload. This would mean the corefx test runner/script would have to change.

On the coreclr side:

* Use the dependency flow/BAR to get a corefx build id in coreclr repo. I’m going to start with a fixed LKG corefx build and hook up the dependency flow stuff later.
* In the coreclr test pipeline, build the URL of the above test asset zips.
* Build the “correlation” payload that contains the coreclr bits to test and the corefx bits from that build id. There are some details here that need to be worked out.
* Submit the Helix job with this URL of the test zips, the correlation payload and a test runner that can download and run the tests. There is already a test runner (Anton wrote) that was written for this that also allowed filtering of tests for “breaking” changes.

**Comment by marek-safar** [comment]
@mikem8361 could the payload url builder and test runner be available in some form of shared/reusable component outside of coreclr?

**Comment by mikem8361** [comment]
Yes, I try to make it shareable/reusable.

**Comment by jkotas** [comment]
I would like the scheme to work well for test asserts of other repos, not just for corefx.

@KevinRansom Do you think the scheme described above (https://github.com/dotnet/corefx/issues/34957#issuecomment-459069106) would work for packing  FSharp xunit tests?

**Comment by jkotas** [comment]
It would be a good idea to talk about requirements of the solution. Here is my take:

## Requirements

- It is easy for any repo with ordinary xunit tests to participate. It should be less than a day of work to onboard a new repo into the system (e.g. to add a option to run the repo tests in CoreCLR CI). We can use FSharp to validate this.

- The underlying storage and versioning scheme is same as what is used to exchange product bits and other binary assets between repos. (e.g. the CoreFX tests are stored in the same places as `Microsoft.Private.CoreFx.NETCoreApp` and have same version)

## Non-goals

- Publish the tests as pre-packaged Helix payloads. It is fine to have a build step that downloads the tests package and repackages it into Helix payloads.

Thoughts?

**Comment by mikem8361** [comment]
I work towards these goals, but I'm only part time on this engineering effort. 

If your requirements are saying that we package the tests in nuget packages instead of zips, I have to disagree.  It is so much easier to create the helix payloads when the tests are zip files. The test asset will be pushed in the same azo blob storage as the product nupkgs and have the same versioning.  The corefx tests and I'm sure any other repo's tests already package the assets up as zips in the build step.  Adding another step to package as nuget doesn't make sense and adds more time to the build. And on the coreclr (and other repos) side that want to run these tests they are going to use Helix so it just make sense to packaged them as Helix playloads.  The nuget package layout and restore doesn't make it easy to use them to run tests in helix.

If we use nupkgs for the test assets this is what would have to happen (roughly):

* Build and zip the test assets. Any repo that runs these tests in Helix will have to do this anyway.  
* Create a nupkg package of all the zipped test assets. It wouldn't make sense to create nupkg(s) of the test assets directly since Helix needs zip files as the payloads. 
* Publish the nupkg to the dotnet-core blob feed.
* Submit the test zips to Helix.

On the coreclr side:

* Install/restore the test asset package(s) somewhere. A download that isn't really needed and takes time.
* Copy the nupkg package layout to something that can be zipped unless the test assets is one nupkg of zip files.
* Create the helix payload which will upload them unnecessarily.

Here is what I plan to do and it should lend itself to any repo and any test assets:

* Build and zip the test assets. Any repo that runs these tests in Helix will have to do this anyway.  
* Publish the zip files to dotnet-core azo container storage.  Part of this is a manifest json file that is generated by the blob feed build task. 
* Submit the test zips to Helix.

On the coreclr side (or any other repo running the tests):

* Download this manifest file and use it to build the helix payload. I plan to add a simple build task to Microsoft.DotNet.Build.Tasks.Feed to parse the manifest file entries and return a ItemGroup that can be used to create the Helix SDK's HelixWorkItems.  The payloads will be uri's of the test asset zip files.  The Helix SDK will need a minor modification to accept a URI PayloadArchive. It already accepts URI correllation payloads. 
* Create the correlation payload combining the corefx bits and the coreclr bits to be tested. There are some details here I need to work out, but this would probably be similar in both scenarios.
* Submit the helix job without actually downloading any of the test assets directly.

This plan involves adding a small build task to Arcade to parse the manifest and a small tweak to the Helix SDK to accept a URI as a payload.

**Comment by jkotas** [comment]
CoreCLR (and I expect Mono too) needs ability to run the tests with different modes. The different modes include setting environment variables, running crossgen on the tests before the run, or running tools like asmdiffs on the output of the tests.

How are such customizations going to be injected into the prebuilt Helix payloads?

**Comment by jkotas** [comment]
> Any repo that runs these tests in Helix will have to do this anyway

What does the adoption of Helix looks like across .NET Core repos? E.g. Is FSharp running its tests in Helix or plans to do so?

**Comment by mikem8361** [comment]
I'll have to think about the details, but the zip files are just the tests themselves and a script to run them.  Setting up the environment, etc. is still part of the step in the repo running the tests.  If the payload itself needs to be customized, I'll have to think about that.  I'm currently trying to get back to what coreclr was doing before the corefx/coreclr shift to Arcade/Azo, etc. which is to just run the tests the way they were run in corefx replacing the coreclr bits. 

If more has to be done to the payload, like running crossgen on the tests, then my approach would require unzip each test zip and rezipping so it may require me to rethink it.  

I assume that Helix is the way to run tests, but I'm sure there is still a lot of work left to get there.  I don't know about any specific repo other than coreclr.

**Comment by safern** [comment]
We could directly change the scripts in corefx to accept optional parameters to enable crossgen or other tools. That would also work. 

Once that is changed, then you wouldn't need to unzip the test zips and you can just add the necessary tools as correlation payload, and pass it in to the script through the optional parameters.

**Comment by jkotas** [comment]
> We could directly change the scripts in corefx to accept optional parameters to enable crossgen or other tools. That would also work.

I think it is important for the script that drive the codegen test modes to live in the CoreCLR repo and not be part of the corefx, FSharp, etc. test drops. Otherwise, it would be very hard to evolve them.

> I'm currently trying to get back to what coreclr was doing

coreclr was doing repackaging before the shift. It did not try to reuse the helix payloads. Do we really believe that skipping the repackaging is that important optimization?

**Comment by marek-safar** [comment]
We are fine with zips (it's actually prefered setup for us over nugets).

>CoreCLR (and I expect Mono too) needs ability to run the tests with different modes. 

I am not sure we want to go there. The CoreFX tests will be in part dynamic and I think they already are. Running the tests with e.g. AOT engine should not be too different than checking for any other feature (e.g. SSL version). Where to put the runtime config options to run them is a good question but I suspect they will be a test configuration specific and therefore part of the test blob to runtime mapping logic.

**Comment by mikem8361** [comment]
Actually coreclr did reuse the helix payload before the new helix SDK changes that made the blob storage internal.  We are currently testing against some very old corefx/corefx tests bits.

The test scripts Santi are referring too are the ones that run the individual test but the main script/build logic to set it all up will still live in the coreclr (or other) repo.  There could be a global script included as part of the correlation payload the script that the individual test script calls to do this kind of setup or initialization.  I don't think coreclr should have knowledge or scripts on how to run an individual test or even the list of tests "hard coded".

**Comment by jkotas** [comment]
> Running the tests with e.g. AOT engine should not be too different than checking for any other feature (e.g. SSL version).

I meant that the test itself is AOT compiled before running.

**Comment by jkotas** [comment]
> coreclr should have knowledge or scripts on how to run an individual test 

What would be the process for investigating a test failure if there is no script to run individual test?

**Comment by marek-safar** [comment]
> I meant that the test itself is AOT compiled before running. That is different from checking SSL version, correct?

Correct, what I meant is the fact that the test is expected to fail or is skipped under AOT or any other runtime combination should be transparent to the launching script.

**Comment by mikem8361** [comment]
There is a script to run the individual test as part of the zip file.  We will eventually have to make it easy to download a specific test locally and run it.  I'm not sure how we run individual corefx tests now in coreclr.  This was originally about run corefx tests in our CI jobs.

In the end, I think my approach can be flexible enough (with Santi's idea of correlation payload script that all the individual scripts run) to handle all of these future requirements.

**Comment by jkotas** [comment]
I have updated the requirements post above https://github.com/dotnet/corefx/issues/34957#issuecomment-459415999 with this discussion. Do they look good?

**Comment by marek-safar** [comment]
@jkotas looks good (I made tiny tweaks) but I am wondering if we should extend this to VM/runtime tests as well.

**Comment by jkotas** [comment]
Yes, the same scheme should be usable to share e.g. CoreCLR tests with Mono. The VM/runtime tests are not ordinary xunit tests so that may need a special treatment in a few places, but most of it should work same.

**Comment by jkotas** [comment]
The helix blobs used for CoreCLR CI just disappeared from the network: https://github.com/dotnet/coreclr/pull/22548#issuecomment-462875777 that make all CoreFX test legs in CoreCLR CI fail.

This is the danger of reusing helix blobs for this that do not have well defined retention policy.

**Comment by ViktorHofer** [comment]
> We could directly change the scripts in corefx to accept optional parameters to enable crossgen or other tools. That would also work.

FWIW the ability to inject a script to the test runner script, i.e. to test under different JIT stress modes, is possible: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.CoreFxTesting/build/core/Core.targets#L107-L114. The problem with that approach is that this has to be injected at build time of the scripts, which happens in corefx when building the test assemblies

What options are needed that require changes in the test run script? Isn't setting env vars enough?

What I also would like to know, how is the runtime chosen when using the dotnet toolchain? Currently test invocations in the test run script look like this:

```
   %RUNTIME_DIR%\dotnet.exe xunit.console.dll System.Text.RegularExpressions.dll [arguments]
```

**Comment by jkotas** [comment]
> What options are needed that require changes in the test run script? Isn't setting env vars enough?

Setting env variables is not enough. We want to be able to do things like compile the test using AOT compiler; and then run it.

> What I also would like to know, how is the runtime chosen when using the dotnet toolchain?

The test harness in the CoreCLR repo knows the path to the runtime that it wants to run the test against.

**Comment by ViktorHofer** [comment]
> The test harness in the CoreCLR repo knows the path to the runtime that it wants to run the test against.

Sounds like you don't use/need the RunScript.bat/sh files that we generate in corefx. Is that correct?

**Comment by jkotas** [comment]
Correct.

Also, we need to be able to amend the xunit command line options with the tests that are failing (e.g. because of CoreCLR and CoreFX are out of sync temporarily). Here is what the building of xunit command line arguments looks like in CoreCLR today:
https://github.com/dotnet/coreclr/blob/master/tests/src/Common/CoreFX/TestFileSetup/Helpers/TestRunHelper.cs#L130

Do all XUnit tests in CoreFX have the same set of command line arguments (traits, etc.)? If there are XUnit assemblies in CoreFX that need an extra command line arguments by default to pass, it would be nice if they can be part of the payload (e.g. in a side-car `<Assembly-Name>.XUnit.rsp` file).

**Comment by ViktorHofer** [comment]
Does it make sense to add switches/tweaks so that both coreclr and mono could use the script that we generate? Adding support for an rsp file in the test directory / some other place is straightforward.

Looking at https://github.com/dotnet/coreclr/blob/master/tests/src/Common/CoreFX/TestFileSetup/Program.cs#L118 it seems like you pass in the path to the dotnet host that you want to use. Why not just binplace the coreclr/mono live build assets into the runtime folder that we send with the payload and use our script instead of writing one and get rid of the whole custom bootstrap logic?

> Do all XUnit tests in CoreFX have the same set of command line arguments (traits, etc.)? If there are XUnit assemblies in CoreFX that need an extra one-off command line arguments by default to pass, it would be nice if they can be part of the payload (e.g. in a side-car `<Assembly-Name>.XUnit.rsp` file).

Yes, fortunately we are very consistent here. The only thing that needs special treatment are runtime settings like the Invariant switch in a runtimeconfig.json file.

**Comment by jkotas** [comment]
What do you mean by the "binplace the coreclr/mono live build assets into the runtime folder that we send with the payload"?

**Comment by ViktorHofer** [comment]
Our runtime folder includes everything (and as you noted, too much) required to run our tests. Assets that corefx doesn't produce live but are included in the runtime directory are either support assemblies like xunit.* or runtime assemblies like coreclr.dll and others. Hence why not overwrite the distributed runtime assets with the live built ones and run the tests the way they are intended to be run?

**Comment by jkotas** [comment]
We need to be able to run the tests against a shipping runtime installed in a read-only global location; without copying random junk to that global location. It is how the tests are meant to be run - everybody else does it that way.

The CoreFX testing has a hole today: As far as I know, we never run the tests against a shipping product today, and we were pretty close to shipping broken bits multiple times because of that. We were lucky that the problem was caught by adhoc testing.

Once we have the test drops, I would like to use them to close this test hole.

**Comment by ViktorHofer** [comment]
> The CoreFX testing has a hole today

Right, which is tracked by https://github.com/dotnet/corefx/issues/33958.

> Once we have the test drops, I would like to use them to close this test hole too.

Got it. To get a bigger picture, please try to answer the open question from above: Could coreclr and mono use the test run script that we generate instead of the custom bootstrapping logic that is currently happening in coreclr? The first argument of the script is the runtime directory, means you can point to that crossgened shipping runtime directory that you wanna use for testing.

**Comment by jkotas** [comment]
> Could coreclr and mono use the test run script that we generate instead of the custom bootstrapping logic that is currently happening in coreclr?

I do not think it is desirable. It is common to customize the test invocation heavily for different stress modes of the runtime. If the corefx test script is used to invoke the tests, there will be trend to move the customization to corefx repo that is a wrong place for it to live.

**Comment by jkotas** [comment]
It is ok with me to include the .cmd to invoke the test in the payload so the places where it works fine can just use it. It is small so it is not a big deal. I am just saying that we should be trying to force all places to use it.

**Comment by ViktorHofer** [comment]
Ok now we are getting to the base. At the end of the day what you need is just the output of all test assemblies (test library, runtimeconfig, test assets) and preferably also the support assemblies to run the tests (xunit)?

**Comment by jkotas** [comment]
That sounds about right.

**Comment by safern** [comment]
Corefx is already publishing from official builds -- @mikem8361 would you mind adding some docs on how to consume this assets for other repos to do so?

**Comment by filipnavara** [comment]
Mono is currently consuming the test assets, but we are running into problem with tests for non-inbox assemblies. Currently the execution environment is basically .NET Core SDK w/ the the downloaded test assets. Thus any non-inbox assemblies, such as `System.Security.Cryptography.Pkcs` or `System.IO.Packaging` are missing and the tests for them fail. These assemblies are not present in `Microsoft.Private.CoreFx.NETCoreApp` NuGet either. How is one supposed to replicate the testhost environment of CoreFX with these bits?

Furthermore, some tests for inbox assemblies indirectly depend on the non-inbox ones. This is the case of `System.Security.Cryptography.Algorithms.Tests` where `CryptoConfig` tests expects to be able to load `System.Security.Cryptography.Pkcs`.

**Comment by ViktorHofer** [comment]
OOB packages are published to the dotnet-core feed. The version matches the test assemblies one (which should be darc updated in the mono repository as commented in one of the PRs in the mono repository). I don't know if we have a list which packages are currently OOB'd. We might want to generate a manifest for that. @safern or @ericstj should know if we already have a manifest.

![image](https://user-images.githubusercontent.com/7412651/55403402-120e8280-5556-11e9-88e3-0e52b62009cc.png)

**Comment by filipnavara** [comment]
I realize that they are publish to the feed, but it's difficult to consume since there's no list of them. I hesitate to hard-code the list since it's bound to change.

**Comment by ViktorHofer** [comment]
I took a quick look, we upload `microsoft.private.packagebaseline` which tells you which packages are inbox and which not. Seems like the data is not up-to-date. AFAIK we have a task in corefx which updates it, let me try that.

[packageIndex.txt](https://github.com/dotnet/corefx/files/3034001/packageIndex.txt)

**Comment by ericstj** [comment]
It’s shouldn’t be out of date, we validate it on build: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Build.Tasks.Packaging/src/ValidateFrameworkPackage.cs.

**Comment by ViktorHofer** [comment]
With out of date I meant that it probably contains many packages that aren't needed and not live built anymore by corefx.
@ericstj does this manifest file also contain entries of other repositories (wpf,wcf,winforms,corclr) or is it  just about corefx?

**Comment by ericstj** [comment]
It's primarily about CoreFx but also contains information about targeting packs and represents assemblies which may not have packages.  Using it as a source for a list of packages isn't really a good idea.  It can be used to help filter a list but shouldn't be the source.

**Comment by ViktorHofer** [comment]
Makes sense. Let's add a task that creates a manifest which is consumeable by Mono.

**Comment by safern** [comment]
Could we publish the testhost? Then mono could basically just copy the .NET SDK  by overriding everything that is inbox, then whatever is OOB will be part of the testhost. Would that work?

**Comment by filipnavara** [comment]
@safern That would actually be easiest to consume by Mono (at least short term).

**Comment by safern** [comment]
> @safern That would actually be easiest to consume by Mono (at least short term).

Great and that should also be an easy fix in the corefx side, it is just adding one more archive to the published list.

**Comment by ViktorHofer** [comment]
I don't think we should do that. The testhost should currently not be consumed by externals as it contains a lot of bloat. Let's plan accordingly first before publishing the testhost.

**Comment by safern** [comment]
> I don't think we should do that. The testhost should currently not be consumed by externals as it contains a lot of bloat. Let's plan accordingly first before publishing the testhost.

I think that is actually what coreclr was planning on doing, replace just the bits needed  in the built testhost, they currently clone corefx and build it to get the testhost... So you would then suggest doing a complicated dance of adding the OOB packages into the manifest (We don't have a tracking list of the OOB packages that are built on the current branch) -- then they would have to add a `PackageReference` to each test project accordingly depending on the OOB package they want to test?

What bloat does it contain other than xunit assemblies?

**Comment by ViktorHofer** [comment]
> then they would have to add a PackageReference to each test project accordingly depending on the OOB package they want to test?

No. The test assemblies already contain the Reference to the OOB packages in the manifest. The OOB assemblies need to be resolvable which means they need to be placed into the runtime directory. The testhost folder in corefx is currently an implementation detail. I fear that as soon as we package it up we are less flexible with changes to it. I would prefer that the OOB packages are consumed via a feed. Having a manifest for all OOB packages shouldn't be too much work. @ericstj opinions?

**Comment by ViktorHofer** [comment]
I'll put this onto our infra board and we will discuss it in our next meeting tomorrow. I understand that this is high prio and will keep you updated.

**Comment by filipnavara** [comment]
> The test assemblies already contain the Reference to the OOB packages in the manifest.

In what manifest exactly? I could be missing something, but I downloaded couple of the .zip test packages and I couldn't find the dependencies listed there.

**Comment by ViktorHofer** [comment]
I was talking about the assembly manifest, i.e. for System.Text.RegularExpressions.Tests:

```
.assembly extern System.Text.RegularExpressions
{
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )                         // .?_....:
  .ver 4:2:1:0
}
.assembly extern System.Diagnostics.Debug
{
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )                         // .?_....:
  .ver 4:1:1:0
}
...
```

**Comment by filipnavara** [comment]
Ah, that. I assumed you meant some manifest like deps.json which would link to NuGet package names.

There are also some tests for inbox assemblies that try load the outbox assemblies dynamically. A good example of that are the test for `System.Security.Cryptography.CryptoConfig`, which tries to load `System.Security.Cryptography.Pkcs.dll`. That cannot be inferred from the manifest alone.

**Comment by ViktorHofer** [comment]
Right, but we don't have a depenedency graph for dynamic dependencies. Usually we use reflection exactly for that reason to not have a dependency from one layer to another. I think having all OOB packages in the runtime directory availble is sufficient, you shouldn't need a more fine-grained dependency manifest, right?

**Comment by ViktorHofer** [comment]
Thanks for letting me know. I think external dependencies should be included in the test output folder, same as we do with xunit. I will add it there for System.Text.Json.

> but it's the only outlier that I know about.

You are right, I just scanned over our runtime references and that's the only external dependencies apart from vstest or xunit dependencies.

**Comment by ViktorHofer** [comment]
@filipvavara one thing that came up during our meeting, is it possible for you guys to restore the OOB packages with the right tfm (netcoreapp) or would you prefer an archive with the actual assemblies?

**Comment by filipnavara** [comment]
Currently the Mono setup mixes downloading the assets through curl, python scripts and NuGet (for Microsoft.Private.CoreFx.NETCoreApp package). Downloading an archive and extracting it is easiest, but NuGet is also an option. I assumed the same packages are going to be consumed by CoreCLR, so that may affect the decision too.

**Comment by ViktorHofer** [comment]
Ok. The reason I'm asking is because we were debating if an archive with just the assemblies would be better than a manifest file which contains entries for the OOB packages and their versions (the versions are different). I would prefer the manifest solution but I'm ok if that's too much overhead.

**Comment by filipnavara** [comment]
At the moment it's easiest in the Mono infrastructure to just download an archive. As @marek-safar previously stated it was the preferred setup over NuGets. I don't want to be short-sighed on this though.

Currently the issue of flowing versions of dependencies from other repositories to Mono is also unsolved problem. If Mono were to use Maestro that would flow the versions and update `MicrosoftPrivateCoreFxNETCoreAppVersion` in Versions.props like in CoreCLR then it would make sense to restore the whole framework package through the same mechanism as in https://github.com/dotnet/coreclr/blob/926b8ac0f0d21d9cdf0b6923a10cc76e099fb495/tests/src/Common/CoreFX/CoreFX.depproj (avoiding listing the individual packages though).

I played with the CoreCLR approach of downloading CoreFX last week but I didn't manage to get it working. That may very much be my fault though. I'll give it another spin to see if I could make it work.

Would it make sense to find a fix for the hard-coded list in CoreCLR as a first step? It would be easier to evaluate whether the same approach would work for Mono.

**Comment by marek-safar** [comment]
@ViktorHofer the archive would work better for us

**Comment by ViktorHofer** [comment]
> I played with the CoreCLR approach of downloading CoreFX last week but I didn't manage to get it working. That may very much be my fault though. I'll give it another spin to see if I could make it work.

Let me know if you need help. I'm here to help you guys onboard and remove any virtual blockers in corefx.

I will try to get this (an archive which contains all OOB assemblies) done this week.

**Comment by filipnavara** [comment]
I've got some temporary solution working in the end by restoring most of the OOB packages through NuGet. One thing that I still didn't fully investigate is that all the tests in `System.ComponentModel.Composition.Registration.Tests` fail with PlatformNotSupportedException. I'll have to check if `System.ComponentModel.Composition.Registration` is supposed to work on macOS or if a wrong configuration of the package is used.

Thanks for all the help!

**Comment by safern** [comment]
It looks like the `System.ComponentModel.Composition.Registration` package only includes a netstandard configuration which just throws PNSE... however in corefx to test it, we have a netcoreapp configuration for the vertical, but that configuarion will not be part of the package and it is just for testing porpuses.

@ViktorHofer I see you added that assembly, is that expected? 

https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.Composition.Registration/src/Configurations.props#L4

**Comment by ViktorHofer** [comment]
Yeah that's expected. We need to wait for netstandard2.1 which will have the necessary APIs to make things work. I recommend to exclude the test assembly for now.

**Comment by ViktorHofer** [comment]
@filipnavara @EgorBo @mikem8361 @adamsitnik https://github.com/dotnet/corefx/pull/37784 adds `Microsoft.Private.CoreFx.OOB`, a meta-package which references all OOB packages. You probably want to use that package instead of manually referencing all OOB packages and setup a darc subscription to it to always stay on the newest version.

**Comment by EgorBo** [comment]
@ViktorHofer not sure I understand how to use it.
I tried to add it to a console app `netcoreapp3.0` with:
```
dotnet publish -f netcoreapp3.0 -r osx-x64 --packages corefx/packages
```

And got:
```
/prj/mono/mono-netcore/netcore/corefx-restore.csproj : error NU1202: Package Microsoft.IO.Redist 4.6.0-preview6.19271.2 is not compatible with netcoreapp3.0 (.NETCoreApp,Version=v3.0). Package Microsoft.IO.Redist 4.6.0-preview6.19271.2 supports: net472 (.NETFramework,Version=v4.7.2)
/prj/mono/mono-netcore/netcore/corefx-restore.csproj : error NU1202: Package Microsoft.IO.Redist 4.6.0-preview6.19271.2 is not compatible with netcoreapp3.0 (.NETCoreApp,Version=v3.0) / osx-x64. Package Microsoft.IO.Redist 4.6.0-preview6.19271.2 supports: net472 (.NETFramework,Version=v4.7.2)
```

**Comment by ViktorHofer** [comment]
That's actually a bug. Will fix that today.

**Comment by ViktorHofer** [comment]
Please try it again with `19271.5`.

**Comment by EgorBo** [comment]
@ViktorHofer yeah, works for me thanks

**Comment by ViktorHofer** [comment]
The remaining work here is factor out non `$(IsNetCoreApp)` assemblies from the shared framework and instead make them app-local so that they are deployed as part of the test app. cc @ericstj
